Subject: [patch 18/43] mv643xx_eth_pcidev: implement hotplug for the marvell gige functionality by probing the northbridge pci id
From: Sven Luther <sl@bplan-gmbh.de>
Patch-mainline: never

Bug 117053 - add mv643xx_eth detection
Bug 161555 - NetworkManager does not work on Pegasos, mv643xx_eth provides no sysfs device symlink

Add Marvell gigabit ethernet hotplug through the northbridge pci id support
for the Pegasos machines.

This patch is not optimal, as the proper way to handling this is to bring
full hotplug support to the platform drivers, but this would be much more
invasive and far reaching a project.  This approach matches the marvell
discovery northbridge pci id to load the driver for the builtin gigabit
ethernet function, and is similar of what is done already using the
discover module loading technique.  With more and more distributions moving
to using hotplug only to probe for modules to load, this patch becomes
necessary.


Not political correct, but we need it anyway because there is no modalias
support for platform devices.

Also, set the /class/net/eth0/device symlink via SET_NETDEV_DEV

Signed-off-by: Olaf Hering <olh@suse.de>

---
 drivers/net/mv643xx_eth.c |   11 +++++++++++
 1 file changed, 11 insertions(+)

Index: linux-2.6.16/drivers/net/mv643xx_eth.c
===================================================================
--- linux-2.6.16.orig/drivers/net/mv643xx_eth.c
+++ linux-2.6.16/drivers/net/mv643xx_eth.c
@@ -39,6 +39,9 @@
 #include <linux/etherdevice.h>
 #include <linux/in.h>
 #include <linux/ip.h>
+#ifdef CONFIG_PPC_CHRP
+#include <linux/pci.h>
+#endif
 
 #include <linux/bitops.h>
 #include <linux/delay.h>
@@ -1445,6 +1448,7 @@ static int mv643xx_eth_probe(struct plat
 		return err;
 	}
 
+	SET_NETDEV_DEV(dev, &pdev->dev);
 	err = register_netdev(dev);
 	if (err)
 		goto out;
@@ -1585,6 +1589,13 @@ MODULE_AUTHOR(	"Rabeeh Khoury, Assaf Hof
 		" and Dale Farnsworth");
 MODULE_DESCRIPTION("Ethernet driver for Marvell MV643XX");
 
+#ifdef CONFIG_PPC_CHRP
+static struct pci_device_id pci_marvell_mv64360[] = {
+	{ PCI_DEVICE(PCI_VENDOR_ID_MARVELL, PCI_DEVICE_ID_MARVELL_MV64360) },
+	{}
+};
+MODULE_DEVICE_TABLE(pci, pci_marvell_mv64360);
+#endif
 /*
  * The second part is the low level driver of the gigE ethernet ports.
  */
